mlflow/tensorflow/__init__.py

```diff
@@ -308,36 +305,94 @@ def save_model(
         mlflow_model.signature = signature
     if input_example is not None:
         _save_example(mlflow_model, input_example, path)
-    root_relative_path = _copy_file_or_tree(src=tf_saved_model_dir, dst=path, dst_dir=None)
-    model_dir_subpath = "tfmodel"
-    model_dir_path = os.path.join(path, model_dir_subpath)
-    shutil.move(os.path.join(path, root_relative_path), model_dir_path)
-
-    flavor_conf = dict(
-        saved_model_dir=model_dir_subpath,
-        meta_graph_tags=tf_meta_graph_tags,
-        signature_def_key=tf_signature_def_key,
-    )
 
-    mlflow_model.add_flavor(FLAVOR_NAME, code=code_dir_subpath, **flavor_conf)
+    if isinstance(model, KerasModel):
+        keras_model_kwargs = keras_model_kwargs or {}
+
+        data_subpath = "data"
+        # construct new data folder in existing path
+        data_path = os.path.join(path, data_subpath)
+        os.makedirs(data_path)
+        model_subpath = os.path.join(data_subpath, _MODEL_SAVE_PATH)
+
+        keras_module = importlib.import_module("tensorflow.keras")
+        # save custom objects if there are custom objects
+        if custom_objects is not None:
+            _save_keras_custom_objects(data_path, custom_objects)
+
+        # save keras module spec to path/data/keras_module.txt
+        with open(os.path.join(data_path, _KERAS_MODULE_SPEC_PATH), "w") as f:
+            f.write(keras_module.__name__)
+
+        # Use the SavedModel format if `save_format` is unspecified
+        save_format = keras_model_kwargs.get("save_format", "tf")
+
+        # save keras save_format to path/data/save_format.txt
+        with open(os.path.join(data_path, _KERAS_SAVE_FORMAT_PATH), "w") as f:
+            f.write(save_format)
+
+        # save keras model
+        # To maintain prior behavior, when the format is HDF5, we save
+        # with the h5 file extension. Otherwise, model_path is a directory
+        # where the saved_model.pb will be stored (for SavedModel format)
+        file_extension = ".h5" if save_format == "h5" else ""
+        model_path = os.path.join(path, model_subpath) + file_extension
+        if path.startswith("/dbfs/"):
+            # The Databricks Filesystem uses a FUSE implementation that does not support
+            # random writes. It causes an error.
+            with tempfile.NamedTemporaryFile(suffix=".h5") as f:
+                model.save(f.name, **keras_model_kwargs)
+                f.flush()  # force flush the data
+                shutil.copyfile(src=f.name, dst=model_path)
+        else:
+            model.save(model_path, **keras_model_kwargs)
+
+        pyfunc_options = {
+            "model_type": _MODEL_TYPE_KERAS,
+            "data": data_subpath,
+        }
```

Comment:
Do we need to include these configs in the pyfunc model? If possible, it would be nice to check in `_load_pyfunc` whether the model has flavor `keras` or `tensorflow` and then read these values from one of those flavors, instead of pyfunc. The MLflow 1.x Keras & TensorFlow flavors don't seem to duplicate this information in both the pyfunc and keras / TF flavor confs.

Reply:
Make sense.